File: plugins/pytest-agent-observability/src/pytest_agent_observability/uploader.py
"""POST the eval payload to agent-observability with retries and local fallback."""

from __future__ import annotations

import json
import logging
import os
import time
from pathlib import Path
from typing import Optional

import httpx


logger = logging.getLogger("pytest_agent_observability")


class UploadConfig:
    __slots__ = ("url", "basic_auth", "timeout_s", "max_retries")

    def __init__(
        self,
        *,
        url: str,
        basic_auth: Optional[tuple[str, str]] = None,
        timeout_s: float = 10.0,
        max_retries: int = 3,
    ) -> None:
        self.url = url.rstrip("/")
        self.basic_auth = basic_auth
        self.timeout_s = timeout_s
        self.max_retries = max_retries


def upload(payload: dict, config: UploadConfig, *, fallback_dir: Optional[Path] = None) -> bool:
    """POST the payload to `{url}/observability/evals/v0`.

    Returns True on success. On total failure (all retries exhausted), writes
    the payload to `fallback_dir/<run_id>.json` if provided and returns False.
    Never raises — the plugin must not break tests.
    """
    endpoint = f"{config.url}/observability/evals/v0"
    auth = config.basic_auth

    last_err: Optional[str] = None
    for attempt in range(1, config.max_retries + 1):
        try:
            with httpx.Client(timeout=config.timeout_s) as client:
                resp = client.post(
                    endpoint,
                    json=payload,
                    auth=auth,
                    headers={"Content-Type": "application/json"},
                )
            if 200 <= resp.status_code < 300:
                return True
            # 4xx is a permanent failure — don't retry.
            if 400 <= resp.status_code < 500:
                last_err = f"HTTP {resp.status_code}: {resp.text[:300]}"
                break
            last_err = f"HTTP {resp.status_code}"
        except Exception as e:
            last_err = f"{type(e).__name__}: {e}"

        if attempt < config.max_retries:
            time.sleep(min(2 ** (attempt - 1), 4))  # 1s, 2s, then 4s

    logger.warning(
        "agent-observability upload failed after %d attempts: %s",
        config.max_retries,
        last_err,
    )
    if fallback_dir is not None:
        _write_fallback(payload, fallback_dir)
    return False


def _write_fallback(payload: dict, fallback_dir: Path) -> None:
    try:
        fallback_dir.mkdir(parents=True, exist_ok=True)
        run_id = payload.get("run", {}).get("run_id", "unknown")
        path = fallback_dir / f"{run_id}.json"
        path.write_text(json.dumps(payload, indent=2))
        logger.warning("wrote eval payload to %s", path)
    except Exception as e:
        logger.error("failed to write fallback payload: %s", e)


def config_from_env() -> Optional[UploadConfig]:
    """Build an UploadConfig from env vars. Returns None if URL is missing.

    Reads `AGENT_OBSERVABILITY_TIMEOUT` (seconds) and
    `AGENT_OBSERVABILITY_MAX_RETRIES` alongside the required URL + optional
    basic-auth pair. Invalid numeric values fall back to the defaults.
    """
    url = os.getenv("AGENT_OBSERVABILITY_URL")
    if not url:
        return None
    user = os.getenv("AGENT_OBSERVABILITY_USER")
    pw = os.getenv("AGENT_OBSERVABILITY_PASS")
    auth: Optional[tuple[str, str]] = (user, pw) if user and pw else None
    return UploadConfig(
        url=url,
        basic_auth=auth,
        timeout_s=_env_float("AGENT_OBSERVABILITY_TIMEOUT", 10.0),
        max_retries=_env_int("AGENT_OBSERVABILITY_MAX_RETRIES", 3),
    )


def _env_float(name: str, default: float) -> float:
    raw = os.getenv(name)
    if raw is None or raw == "":
        return default
    try:
        return float(raw)
    except ValueError:
        return default


def _env_int(name: str, default: int) -> int:
    raw = os.getenv(name)
    if raw is None or raw == "":
        return default
    try:
        return int(raw)
    except ValueError:
        return default

File: plugins/agent-observability-sdk-node/src/livekit/vitest/uploader.ts
import { mkdir, writeFile } from "node:fs/promises";
import path from "node:path";
import type { EvalPayloadV0 } from "./types.js";

export interface UploadConfig {
  url: string;
  basicAuth?: { user: string; pass: string } | null;
  timeoutMs: number;
  maxRetries: number;
}

/**
 * POST the payload to `{url}/observability/evals/v0`.
 * Returns true on success. Never throws.
 */
export async function upload(
  payload: EvalPayloadV0,
  config: UploadConfig,
  opts: { fallbackDir?: string | null; logger?: Logger } = {},
): Promise<boolean> {
  const { fallbackDir, logger = defaultLogger } = opts;
  const endpoint = `${stripTrailingSlash(config.url)}/observability/evals/v0`;

  const headers: Record<string, string> = { "Content-Type": "application/json" };
  if (config.basicAuth) {
    const token = Buffer.from(
      `${config.basicAuth.user}:${config.basicAuth.pass}`,
      "utf8",
    ).toString("base64");
    headers["Authorization"] = `Basic ${token}`;
  }

  const body = JSON.stringify(payload);
  let lastErr = "";

  for (let attempt = 1; attempt <= config.maxRetries; attempt++) {
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), config.timeoutMs);
    try {
      const res = await fetch(endpoint, {
        method: "POST",
        headers,
        body,
        signal: controller.signal,
      });
      if (res.ok) return true;
      // 4xx = permanent; stop retrying.
      if (res.status >= 400 && res.status < 500) {
        lastErr = `HTTP ${res.status}: ${await safeText(res)}`;
        break;
      }
      lastErr = `HTTP ${res.status}`;
    } catch (e) {
      lastErr = `${(e as Error).name}: ${(e as Error).message}`;
    } finally {
      clearTimeout(timer);
    }

    if (attempt < config.maxRetries) {
      await sleep(Math.min(2 ** (attempt - 1), 4) * 1000);
    }
  }

  logger.warn(
    `agent-observability upload failed after ${config.maxRetries} attempts: ${lastErr}`,
  );
  if (fallbackDir) await writeFallback(payload, fallbackDir, logger);
  return false;
}

async function writeFallback(
  payload: EvalPayloadV0,
  dir: string,
  logger: Logger,
): Promise<void> {
  try {
    await mkdir(dir, { recursive: true });
    const runId = payload.run?.run_id || "unknown";
    const p = path.join(dir, `${runId}.json`);
    await writeFile(p, JSON.stringify(payload, null, 2));
    logger.warn(`wrote eval payload to ${p}`);
  } catch (e) {
    logger.error(`failed to write fallback payload: ${(e as Error).message}`);
  }
}

async function safeText(res: Response): Promise<string> {
  try {
    return (await res.text()).slice(0, 300);
  } catch {
    return "";
  }
}

function stripTrailingSlash(s: string): string {
  return s.endsWith("/") ? s.slice(0, -1) : s;
}

function sleep(ms: number): Promise<void> {
  return new Promise((r) => setTimeout(r, ms));
}

export interface Logger {
  info?(msg: string): void;
  warn(msg: string): void;
  error(msg: string): void;
}

export const defaultLogger: Logger = {
  info: (m) => console.log(`[agent-observability] ${m}`),
  warn: (m) => console.warn(`[agent-observability] ${m}`),
  error: (m) => console.error(`[agent-observability] ${m}`),
};

export function configFromEnv(env: NodeJS.ProcessEnv = process.env): UploadConfig | null {
  const url = env.AGENT_OBSERVABILITY_URL;
  if (!url) return null;
  const user = env.AGENT_OBSERVABILITY_USER;
  const pass = env.AGENT_OBSERVABILITY_PASS;
  return {
    url,
    basicAuth: user && pass ? { user, pass } : null,
    timeoutMs: 10_000,
    maxRetries: 3,
  };
}
